File: 17.wip.hs
import Data.Map qualified as M
import Data.Set qualified as S
import Data.Maybe (fromJust, fromMaybe, mapMaybe, listToMaybe, isJust)
import Data.List (find, minimumBy)

main :: IO ()
main = interact $ unlines . (\grid -> concat [p1 grid, p2 grid]) . parse

type Node = (Int, Int)
data Grid a = Grid { items :: M.Map Node a, lastNode :: Node } deriving Show

data Direction = L | R | U | D deriving (Show, Eq, Ord)
data Cell = Cell {
  node :: Node, direction :: Direction,
  -- The number of blocks that we have already moved in this direction.
  moves :: Int }
  deriving  (Show, Eq, Ord)

data Neighbour = Neighbour { cell :: Cell, distance :: Int } deriving Show

parse :: String -> Grid Int
parse s = Grid { items = M.fromList xs, lastNode = fst (last xs) }
  where xs = [((x, y), read [c]) | (y, l) <- enum (lines s), (x, c) <- enum l]

enum :: [a] -> [(Int, a)]
enum = zip [0..]

neighbours :: Grid Int -> [Int] -> Cell -> [Neighbour]
neighbours Grid { items } range = adjacent
  where
    toNeighbour :: Cell -> Int -> [Neighbour] -> (Int, [Neighbour])
    toNeighbour cell d xs = case M.lookup (node cell) items of
      Just d2 | moves cell `elem` range -> (d + d2, Neighbour cell (d + d2) : xs)
      Just d2 -> (d + d2, xs)
      _ -> (d, xs)
    adjacent :: Cell -> [Neighbour]
    adjacent Cell { node = (x, y), direction, moves } =
      let rng = [1..maximum range] in case direction of
        L -> concat [
          snd (foldl (\(d, xs) m -> let cell = Cell (x + m, y) L (moves + m)
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y - m) U m
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y + m) D m
                                    in toNeighbour cell d xs) (0, []) rng)
          ]

        R -> concat [
          snd (foldl (\(d, xs) m -> let cell = Cell (x - m, y) R (moves + m)
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y - m) U m
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y + m) D m
                                    in toNeighbour cell d xs) (0, []) rng)
          ]

        U -> concat [
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y - m) U (moves + m)
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x - m, y) R m
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x + m, y) L m
                                    in toNeighbour cell d xs) (0, []) rng)
          ]

        D -> concat [
          snd (foldl (\(d, xs) m -> let cell = Cell (x, y + m) D (moves + m)
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x - m, y) R m
                                    in toNeighbour cell d xs) (0, []) rng),
          snd (foldl (\(d, xs) m -> let cell = Cell (x + m, y) L m
                                    in toNeighbour cell d xs) (0, []) rng)
          ]

-- Find the shortest path from start to an end using Dijkstra's algorithm.
dijkstra :: Grid Int -> Node -> (Cell -> Bool)
            -> [Int]
            -> (Maybe Int, [String])
dijkstra grid@Grid { items } start isEnd range =
  go (M.singleton startCell 0) M.empty S.empty (h_singleton (0, startCell))
  where
    -- Since moves is 0, the direction doesn't distinguish between moving ahead
    -- or turning.
    startCell = Cell { node = start, direction = L, moves = 0 }
    go :: M.Map Cell Int -> M.Map Cell Cell -> S.Set Cell ->  Heap (Int, Cell) -> (Maybe Int, [String])
    go ds parent seen q = case extractMin q of
        Nothing -> case nearestEnd ds of
                     Nothing -> (Nothing, [])
                     Just (u, du) -> (Just du, showDistanceMap grid ds parent u range)
        Just ((du, u), q')
          | u `S.member` seen -> go ds parent seen q'
          | otherwise ->
             let adj = (neighbours grid range u)
                 adj' = filter (\Neighbour {cell} -> cell `S.notMember` seen) adj
                 (ds', parent', q'') = foldl (relax u du) (ds, parent, q') adj
                 (d', vs) = go ds' parent' (S.insert u seen) q''
             in (d', vs)

    relax :: Cell -> Int -> (M.Map Cell Int, M.Map Cell Cell, Heap (Int, Cell))
             -> Neighbour -> (M.Map Cell Int, M.Map Cell Cell, Heap (Int, Cell))
    relax u du (ds, parent, q) Neighbour { cell = v, distance = d } =
      case M.lookup v ds of
        Just dv | dv < du + d -> (ds, parent, q)
        _ -> (M.insert v (du + d) ds, M.insert v u parent, h_insert (du + d, v) q)

    nearestEnd ds = case map (\k -> (k, fromJust (M.lookup k ds))) $ filter (\k -> isEnd k) (M.keys ds) of
      [] -> Nothing
      kvs -> Just $ minimumBy (\(k, v) (k2, v2) -> v `compare` v2) kvs


data Heap a = Empty | Heap a (Heap a) (Heap a)

union :: Ord a => Heap a -> Heap a -> Heap a
union Empty h = h
union h Empty = h
union hl@(Heap l ll lr) hr@(Heap r _ _)
  | l <= r = Heap l (union hr lr) ll
  | otherwise = union hr hl

extractMin :: Ord a => Heap a -> Maybe (a, Heap a)
extractMin Empty = Nothing
extractMin (Heap x l r) = Just (x, union l r)

h_singleton :: a -> Heap a
h_singleton x = Heap x Empty Empty

h_insert :: Ord a => a -> Heap a -> Heap a
h_insert x h = h_singleton x `union` h

showDistanceMap :: Grid a -> M.Map Cell Int -> M.Map Cell Cell -> Cell -> [Int] -> [String]
showDistanceMap Grid { lastNode = (mx, my) } ds parent end range = map line [0..my]
  where
    path = retrace S.empty end
      where retrace s n = let s' = S.insert n s in case M.lookup n parent of
              Nothing -> s'
              Just p -> retrace s' p
    isOnPath cell = S.member cell path
    line y = unwords $ map dist [0..mx]
      where dist x = showCell $ find isOnPath [
              Cell {node = (x, y), direction = d, moves }
              | d <- [L, R, U, D], moves <- range]
            showCell Nothing = "   .   "
            showCell (Just cell@Cell { node, moves }) =
              " " ++ d ++ " " ++ show moves ++ " "
                where d = pad3 $ show $ fromJust $ M.lookup cell ds
                      pad3 s = reverse $ take 3 (reverse ("   " ++ s))

p1, p2 :: Grid Int -> [String]
p1 grid = runP grid [1..3]
p2 grid = runP grid [4..10]

runP :: Grid Int -> [Int] -> [String]
runP grid range = let (r, zs) = dijkstra grid (0, 0) isEnd range
          in zs ++ ["shortest-path result " ++ (show $ fromMaybe (-1) r)]
  where
    isEnd Cell { node } = node == (lastNode grid)

File: 17.wip.swift
struct Grid {
    let items: [Node: Int]
    let lastNode: Node
}

struct Node: Hashable {
    let x, y : Int
}

func readInput() -> Grid {
    var items = [Node: Int]()
    var y = 0, x = 0
    while let line = readLine() {
        x = 0
        for c in line {
            items[Node(x: x, y: y)] = c.wholeNumberValue!
            x += 1
        }
        y += 1
    }
    return Grid(items: items, lastNode: .init(x: x - 1, y: y - 1))
}

enum Direction {
    case l, r, u, d
}

struct Cell: Hashable {
    let node: Node
    let direction: Direction
    let moves: Int
}

func shortestPath(grid: Grid, moveRange: ClosedRange<Int>) -> Int {
    let startNode = Node(x: 0, y: 0)
    // Setting moves to 0 to allows us to equally consider both left and down
    // neighbours.
    let startCell = Cell(node: startNode, direction: .l, moves: 0)
    func isEnd(_ cell: Cell) -> Bool { cell.node == grid.lastNode }

    func adj(_ u: Cell) -> [Neighbour] {
        neighbours(grid: grid, moveRange: moveRange, cell: u)
    }

    var dist: [Cell: Int] = [startCell: 0]
    var seen: Set<Cell> = Set()
    //var invDist: [Int: [Node]] = []
    //var minDist = 0

    func popNearest() -> (Cell, Int)? {
        var u: Cell?
        var du: Int = .max
        for (v, dv) in dist {
            if dv < du && !seen.contains(v) {
                u = v
                du = dv
            }
        }
        if let u { return (u, du) }
        return nil
    }

    while let (u, du) = popNearest() {
        if !seen.insert(u).inserted { continue }
        if isEnd(u) { return du }
        for n in adj(u) {
            let v = n.cell
            let d = n.d
            let dv = dist[v] ?? .max
            if du + d < dv {
                dist[v] = du + d
            }
        }
    }

    return 0
}

struct Neighbour {
    let cell: Cell
    let d: Int
}

func neighbours(grid: Grid, moveRange: ClosedRange<Int>, cell: Cell) -> [Neighbour] {
    let node = cell.node
    let moves = cell.moves
    let (x, y) = (node.x, node.y)
    let nbrs: [[Neighbour]]

    func make(_ xy: (Int, Int), _ direction: Direction, _ moves: Int) -> Cell {
        Cell(node: Node(x: xy.0, y: xy.1), direction: direction, moves: moves)
    }

    let cellRange = 1...moveRange.upperBound

    func seq(_ makeCell: (Int) -> Cell) -> [Neighbour] {
        var s = 0
        return Array(cellRange.compactMap { m in
            let cell = makeCell(m)
            guard let d = grid.items[cell.node] else { return nil }
            s += d
            if moveRange.contains(cell.moves) {
                return Neighbour(cell: cell, d: s)
            }
            return nil
        })
    }

    switch cell.direction {
    case .l:
        nbrs = [seq({ m in make((x + m, y), .l, moves + m) }),
                seq({ m in make((x, y - m), .u, m) }),
                seq({ m in make((x, y + m), .d, m) })]
    case .r:
        nbrs = [seq({ m in make((x - m, y), .r, moves + m) }),
                seq({ m in make((x, y - m), .u, m) }),
                seq({ m in make((x, y + m), .d, m) })]
    case .u:
        nbrs = [seq({ m in make((x, y - m), .u, moves + m) }),
                seq({ m in make((x - m, y), .r, m) }),
                seq({ m in make((x + m, y), .l, m) })]
    case .d:
        nbrs = [seq({ m in make((x, y + m), .d, moves + m) }),
                seq({ m in make((x - m, y), .r, m) }),
                seq({ m in make((x + m, y), .l, m) })]
    }
    return nbrs.flatMap { $0 }
}

let grid = readInput()
let p1 = shortestPath(grid: grid, moveRange: 1...3)
let p2 = shortestPath(grid: grid, moveRange: 4...10)
print(p1, p2)
